File: project1/src/task1.py
import json
import logging
import os
import re
import string
import sys
from datetime import datetime

from pyspark import SparkConf, SparkContext
import time

stopwords = None

def isValidFiles(input_file, stop_words_file):
    if not os.path.isfile(input_file) or not os.path.isfile(stop_words_file):
        return False
    return True

def mapJsonObj(jsonObj):
    year = datetime.strptime(jsonObj["date"], "%Y-%m-%d %H:%M:%S").year
    return ((year, 1), (jsonObj["user_id"], 1), jsonObj["text"])

def loadStopWords(file_path):
    data = None
    with open(file_path, "rb") as file:
        data = file.read()
    return data.decode("utf-8").split()

def filterPunctuation(text):
    words = text.translate(str.maketrans('', '', string.punctuation))
    return words.split()

def filterWords(word):
    pattern = "[a-zA-Z]+"
    if re.match(pattern, word) and word not in stopwords:
        return True
    return False

def sort_values(results):
    temp_dict = {}

    for pair in results:
        if temp_dict.get(pair[1]) == None:
            temp_dict[pair[1]] = [pair[0]]
        else:
            temp_dict[pair[1]].append([pair[0]])

    sorted_val = []

    for k, values in temp_dict.items():
        values = sorted(values)

        for value in values:
            sorted_val.append([value, k])

    return sorted_val

def main(argv):
    global stopwords

    assert len(argv) == 6, "Script takes 6 arguments"

    input_file, output_file, stop_words_file, y_arg, m_arg, n_arg = argv

    if not isValidFiles(input_file, stop_words_file):
        print("Invalid file path argument")
        return

    y_arg = int(y_arg)
    m_arg = int(m_arg)
    n_arg = int(n_arg)

    start = time.time()

    out_dict = {}

    # Get Error Logs
    logging.getLogger("org").setLevel(logging.ERROR)

    conf = SparkConf().setMaster("local[*]").setAppName("Task1")
    sc = SparkContext(conf=conf).getOrCreate()

    lines = sc.textFile(input_file)
    rdd = lines.map(json.loads) \
               .map(lambda x: mapJsonObj(x)).cache()
    
    # Determine the reviews in a given year
    reviews_y = rdd.filter(lambda x: x[0][0] == y_arg) \
                   .map(lambda x: x[0]) \
                   .reduceByKey(lambda x, y: x + y)

    # Determine the top m users who have the larges number of reviews and their count
    user_reviews = rdd.map(lambda x: x[1]) \
                      .reduceByKey(lambda x, y: x + y) \
                      .sortBy(lambda x: x[1], ascending=False)

    total_reviews = rdd.count()
    unique_users = rdd.map(lambda x: x[1][0]).distinct().count()
    top_users = user_reviews.take(m_arg)

    stopwords = loadStopWords(stop_words_file)

    text_rdd = rdd.flatMap(lambda x: filterPunctuation(x[2])) \
                         .map(lambda x: (x.lower(), 1)) \
                         .filter(lambda x: filterWords(x[0])).cache()


    top_words = text_rdd.reduceByKey(lambda x, y: x + y, 8) \
                        .sortBy(lambda x: x[1], ascending=False)

    # Write output
    out_dict["A"] = total_reviews
    out_dict["B"] = reviews_y.collect()[0][1]
    out_dict["C"] = unique_users
    out_dict["D"] = sort_values(top_users)

    top_n_words = top_words.take(n_arg)
    top_n_words_list = [word[0] for word in top_n_words]

    out_dict["E"] = sorted(top_n_words_list)

    end = time.time()

    print("Taks 1", end - start)


    with open(output_file, "w") as file:
        json.dump(out_dict, file)

if __name__ == "__main__":
    main(sys.argv[1:])

File: project1_scala/hw1/src/main/scala/task1.scala
import java.io.FileWriter
import java.nio.charset.StandardCharsets
import java.nio.file.{FileSystemNotFoundException, Files, Paths}

import org.apache.log4j.{Level, Logger}
import org.apache.spark.SparkContext
import org.json4s._
import org.json4s.jackson.JsonMethods._
import org.json4s.JsonDSL._

import scala.collection.mutable.ListBuffer
import scala.util.Sorting.quickSort
import Array._
import scala.collection.immutable.ListMap

//noinspection DuplicatedCode
object task1 extends Serializable {

  def time[R](block: => R): R = {
    val t0 = System.nanoTime()
    val result = block
    val t1 = System.nanoTime()
    println("Elapsed time: " + (t1 - t0)/1e9d + "sec")
    result
  }

  def loadStopWords(stopWordsFile: String): Option[Array[String]] = {
    try {
      val bytesArray = Files.readAllBytes(Paths.get(stopWordsFile))
      val resultString = new String(bytesArray, StandardCharsets.UTF_8)
      Some(resultString.split("\\s+"))
    } catch {
      case x: FileSystemNotFoundException => {
        println("Invalid file path for Stop words")
        None
      }
    }
  }

  type reviewsType = ((Int, Int), (String, Int), String)
  def mapJsonObj(obj: JObject): reviewsType = {
    implicit val formats = DefaultFormats

    var dateString = (obj \ "date").extract[String]
    var year: Int = dateString.split("-")(0).toInt

    var userId: String = (obj \ "user_id").extract[String]

    var text: String = (obj \ "text").extract[String]

    ((year, 1), (userId, 1), text)
  }

  def filterPunctuation(str: String): Array[String] = {
    val words = str.replaceAll("\\p{Punct}", "")
    words.split("\\s+")
  }

  def sortValues(tuples: Array[(String, Int)]): String = {
    var tempMap = Map[Int, Array[String]]()

    for (i <- tuples) {
      if (!(tempMap isDefinedAt i._2)) {
        tempMap += (i._2 -> Array(i._1))
      } else {
        val tempArr: Array[String] = tempMap(i._2)
        val tempBuffer = Array(i._1)
        tempMap += (i._2 -> concat(tempArr, tempBuffer))
      }
    }

    //Resort map by key
    val tempMapSorted = ListMap(tempMap.toSeq.sortWith(_._1 > _._1):_*)

    var result = "["

    for ((k, v) <- tempMapSorted) {
      quickSort(v) //sort array
      for (i <- v) {
        result += "[\"" + i + "\", " + k + "], "
      }
    }
    result = result.substring(0, result.length - 2)
    result += "]"
    result
  }

  def getResults(inputFile: String, stopWordsFile: String, y: Int, m: Int, n: Int): String = {
    Logger.getLogger("org").setLevel(Level.ERROR)

    val sc = new SparkContext("local[*]", "Task1")

    val lines = sc.textFile(inputFile)
    val rdd = lines.map(x => parse(x).asInstanceOf[JObject]).map(x => mapJsonObj(x)).cache()

    // Determine the reviews in a given year
    val reviewsYear = rdd.filter(x => x._1._1 == y)
      .map(x => x._1).
      reduceByKey((x, y) => x + y)

    // Determine the top m users who have large number of reviews and their count
    val userReviews = rdd.map(x => x._2)
      .reduceByKey((x, y) => x + y)
      .sortBy(_._2, ascending = false)

    val totalReviews = rdd.count()
    val uniqueUsers = rdd.map(x => x._2._1).distinct().count()
    val topUsers = userReviews.take(m)

    // Load stopWords
    val stopWords = loadStopWords(stopWordsFile).getOrElse(Array[String]())

    val textRDD = rdd.flatMap(x => filterPunctuation(x._3))
      .map(x => (x.toLowerCase(), 1))
      .filter(x => x._1.matches("[a-zA-Z]+") && (stopWords.indexOf(x._1) == -1)).cache()


    val topWordsRDD = textRDD.reduceByKey((x, y) => x + y, 8)
      .sortBy(_._2, ascending = false)

    val topWords = topWordsRDD.take(n)
    val topWordsListBuffer: ListBuffer[String] = ListBuffer()

    for (i <- topWords) {
      topWordsListBuffer += i._1
    }

    var topWordsList = topWordsListBuffer.toList
    topWordsList = topWordsList.sorted


    val out_dict = ("A" -> totalReviews) ~ ("B" -> reviewsYear.collect()(0)._2) ~
      ("C" -> uniqueUsers) ~ ("D" -> parse(sortValues(topUsers))) ~ ("E" -> topWordsList)

    compact(render(out_dict))
  }
  def main(args: Array[String]): Unit = {
    // Read input arguments
    val Array(inputFile, outputFile, stopWordsFile, y, m, n) = args

    val result = time(getResults(inputFile, stopWordsFile, y.toInt, m.toInt, n.toInt))

    val writer = new FileWriter(outputFile)
    writer.write(result)
    writer.close()
  }
}
